### PR TITLE
add minimal PackageManager and Lockfile interfaces

### PR DESCRIPTION
**Pull request by tylerd-canva**
## Hey, I just made a Pull Request!

This is a smaller, independent changes split out from https://github.com/backstage/backstage/pull/27095. This is working toward PNPM support as mentioned in https://github.com/backstage/backstage/issues/20658.

I have only added the minimal Yarn implementation for now as well as some supporting changes in the run/errors lib files. Nothing new is being used or exported anywhere.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/cli-node/src/pacman/yarn/Yarn.ts
Comment (tylerd-canva) [inline packages/cli-node/src/pacman/yarn/Yarn.ts:94]: This has been taken from https://github.com/backstage/backstage/blob/a7e82c9b017424f4adeab90fd3962b11fc788979/packages/cli/src/lib/yarn.ts#L25. It will be deduplicated in a later change.

File: packages/cli-node/src/run.ts
Comment (tylerd-canva) [inline packages/cli-node/src/run.ts:1]: The new code in this file has been taken from https://github.com/backstage/backstage/blob/9f9cad3aff39cfa16d7eeb4819a14d4f2a6588cf/packages/cli/src/lib/run.ts and will be deduplicated in a later PR.

Comment (tylerd-canva) [inline packages/cli-node/src/run.ts:1]: The new code in this file has been taken from https://github.com/backstage/backstage/blob/9f9cad3aff39cfa16d7eeb4819a14d4f2a6588cf/packages/cli/src/lib/run.ts and will be deduplicated in a later PR.

File: packages/cli-node/src/pacman/PackageManager.ts
Comment (tylerd-canva) [reply packages/cli-node/src/pacman/PackageManager.ts:56]: Thanks for bringing up this use case. How does it currently work for you with Yarn? Do you have a `yarn.lock` and `package.json` (with the `workspaces` field) at the root of your Backstage instance (rather than the monorepo root) and does `backstage-cli` correctly locate these files?

In my test branch with PNPM (not pushed yet) I've been looking to replicate the same behaviour, but it sounds like we might need additional configuration in your case since the PNPM files won't be located in the same place the Yarn files are.

Comment (aramissennyeydd) [reply packages/cli-node/src/pacman/PackageManager.ts:56]: We run the base Yarn setup in a separate monorepo, this work would help us move to an existing Rush monorepo. Maybe this would fit into `backstage.json`? Rush has this idea of a packageManager field in `rush.json`, https://rushjs.io/pages/maintainer/package_managers/#specifying-your-package-manager. And we could allow additional config for where your lockfile is etc

Comment (tylerd-canva) [reply packages/cli-node/src/pacman/PackageManager.ts:56]: Very interesting. Rush looks pretty cool and I had not heard of it so thanks again for bringing this up. I'm not sure what the right path forward here is, it almost seems like Rush could be it's own implementation of `PackageManager`/`Lockfile` but then again it still uses Yarn/pnpm/npm under the hood so there'd be a lot of shared code. Probably the simplest option would be to support Rush similarly to how lerna is supported, with a few branching conditions to retrieve files from the right locations where needed. Unfortunately I don't think it's a good idea to include that scope in these changes right off the bat, but it will certainly be much easier to do once we have the base Pnpm support merged.

Comment (aramissennyeydd) [reply packages/cli-node/src/pacman/PackageManager.ts:56]: Happy to have this be handled later - just wanted to bring it up a little early. Basically, the biggest thing for me is that I'd love to see package manager be definable and lockfile location be customizable, yarn/npm/pnpm. I can run with the logic from there. A lot of libraries try the autodiscovery method and it falls down for Rush 😕.

**Comment by jhaals** [comment]
Thanks for breaking this out into a smaller PR 👍 Will loop @Rugvip in on this one

**Comment by tylerd-canva** [comment]
@Rugvip -- sorry to ping, I'm sure you're very busy, but I'm just mindful of losing context on this one as I work on other things. Any thoughts/feedback would be appreciated.

File: packages/cli-node/src/pacman/PackageManager.ts
Comment (paulius-valiunas) [inline packages/cli-node/src/pacman/PackageManager.ts:44]: these could be changed to getters for convenience

File: packages/cli-node/src/pacman/PackageManager.ts
Comment (paulius-valiunas) [reply packages/cli-node/src/pacman/PackageManager.ts:56]: rush "looks" cool, but it's not very useful. I feel that it lacks behind in features compared to a plain pnpm monorepo and slows us down more than anything. I've had incomparably better experience with pnpm+turbo+beachball (or changesets). Rush is a jack of all trades, master of none.

**Review by Rugvip — APPROVED** [review]
Thank you! 👍 

Yep sorry been out for KubeCon, and queue was quite long already 😅

I'm actually gonna go ahead and merge as is and we can hopefully continue improving it before the 1.34 release next month. Won't treat these as very stable interfaces to start with either.

We could do an `/alpha` export but let's not bother with that for now

**Comment by Rugvip** [comment]
@tylerd-canva just needs a rebase to bring in the node 22 workflows and resolve the conflict

**Pull request merged**
